src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,973 @@
+                                    (err ERR_PANIC)))))
+            (asserts!
+              can-new-owner-get-name
+              (err ERR_PRINCIPAL_ALREADY_ASSOCIATED))
+            (unwrap!
+              (update-name-ownership? namespace name contract-caller owner-unwrapped)
+              (err ERR_NAME_COULD_NOT_BE_TRANSFERED))
+            (ok true)))))
+    ;; Update the zonefile, if any.
+    (if (is-none zonefile-hash)
+      (map-set name-properties
+        ((namespace namespace) (name name))
+        ((registered-at (some block-height))
+         (imported-at none)
+         (revoked-at none)
+         (zonefile-hash (get zonefile-hash name-props))))
+      (update-zonefile-and-props
+              namespace 
+              name
+              (some block-height)
+              none
+              none
+              (unwrap! zonefile-hash (err ERR_PANIC))))  
+    (ok true)))
+
+;; Additionals public methods
+
+(define-read-only (can-name-be-registered (namespace (buff 19)) (name (buff 16)))
+  (let (
+      (wrapped-name-props (map-get? name-properties ((namespace namespace) (name name))))
+      (current-owner (map-get? owner-name ((owner contract-caller))))
+      (namespace-props (unwrap! (map-get? namespaces ((namespace namespace))) (ok false))))
+    ;; Ensure that namespace has been launched 
+    (unwrap! (get launched-at namespace-props) (ok false))
+    ;; Early return - Name has never be minted
+    (asserts! (is-none (nft-get-owner? names (tuple (name name) (namespace namespace)))) (ok true))
+    ;; Integrity check - Ensure that we have some entries in nft && owner-name && name-props
+    (asserts! 
+      (and 
+        (is-some wrapped-name-props)
+        (is-some current-owner))
+      (err ERR_PANIC))
+    (let ((name-props (unwrap! wrapped-name-props (err ERR_PANIC))))
+      ;; Early return - Name has been revoked and can be registered
+      (asserts! (is-some (get revoked-at name-props)) (ok true))
+      ;; Integrity check - Ensure that the name was either "imported" or "registered".
+      (asserts! 
+        (or 
+          (and (is-some (get registered-at name-props)) (is-none (get imported-at name-props)))
+          (and (is-some (get imported-at name-props)) (is-none (get registered-at name-props))))
+        (err ERR_PANIC))
+      ;; Is lease expired?
+      (is-name-lease-expired? namespace name))))
+
+(define-read-only (name-resolve (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (owner (unwrap!
+      (nft-get-owner? names (tuple (name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND))) ;; The name must exist
+    (name-props (unwrap!
+      (map-get? name-properties ((name name) (namespace namespace)))
+      (err ERR_NAME_NOT_FOUND)))
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (is-lease-expired (is-name-lease-expired? namespace name)))
+    ;; The namespace must be launched
```

Comment:
I think you will want to drop this check -- in Stacks 1.0, a name is resolvable even if it is only imported.  This is because the namespace-reveal principal can spend a long period of time populating the namespace before launching it (i.e. because they're bootstrapping it).  For example, when we launched Stacks 1.0, it took about 3-4 _months_ before we were comfortable launching `.id`.  In the mean time, we simply used `name-import` to update and transfer names on the users' behalf while we were migrating names from Namecoin, and our resolver still resolved imported names (and only fell back to Namecoin if the name hadn't been imported yet).

Reply:
Addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/58c72297359afda14cac67254ae854c93387a583